slycot/exceptions.py

```diff
@@ -0,0 +1,228 @@
+                        break  # docstring
+    except StopIteration:
+        pass
+    return (slycot_error, message)
+
+
+def raise_if_slycot_error(info, arg_list=None, docstring=None, checkvars={}):
+    """Raise exceptions or warnings if slycot info returned is non-zero.
+
+    Parameters
+    ----------
+    info: int
+        The parameter INFO returned by the SLICOT subroutine
+    arg_list: list of str, optional
+        A list of arguments (possibly hidden by the wrapper) of the SLICOT
+        subroutine
+    docstring: str, optional
+        The docstring of the Slycot function
+    checkvars: dict, optional
+        dict of variables for evaluation of <infospec> and formatting the
+        exception message
+
+    Notes
+    -----
+    If the numpydoc compliant docstring has a "Raises" section with one or
+    multiple definition terms ``SlycotError : e`` or a subclass of it,
+    the matching exception text is used.
+
+    To raise warnings, define a "Warns" section using a ``SlycotWarning : e``
+    definition or a subclass of it.
+
+    The definition body must contain a reST compliant field list with
+    ':<infospec>:' as field name, where <infospec> specifies the valid values
+    for `e.ìnfo` in a python parseable expression using the variables provided
+    in `checkvars`. A single " = " is treated as " == ".
+
+    The body of the field list contains the exception or warning message and
+    can contain replacement fields in format string syntax using the variables
+    in `checkvars`.
+
+    For negative info, the argument as indicated in arg_list was erroneous and
+    a generic SlycotParameterError is raised if matching infospec was defined.
+
+    Example
+    -------
+    >>> def fun(info):
+    ...     '''Example function
+    ...
+    ...     Raises
+    ...     ------
+    ...     SlycotArithmeticError : e
+    ...         :e.info = 1: Info is 1
+    ...         :e.info > 1 and e.info < n:
+    ...             Info is {e.info}, which is between 1 and {n}
+    ...         :n <= e.info < m:
+    ...             {e.info} is in [{n}, {m:10.2g})!
+    ...
+    ...     Warns
+    ...     -----
+    ...     SlycotResultWarning : e
+    ...         :e.info >= 120: {e.info} is too large
+    ...     '''
+    ...     n, m = 4, 120.
+    ...     raise_if_slycot_error(info,
+    ...                           arg_list=["a", "b", "c"],
+    ...                           docstring=fun.__doc__,
+    ...                           checkvars=locals())
+    ...
+    >>> fun(0)
+    >>> fun(-1)
+    SlycotParameterError:
+    The following argument had an illegal value: a
+    >>> fun(1)
+    SlycotArithmeticError:
+    Info is 1
+    >>> fun(2)
+    SlycotArithmeticError:
+    Info is 2, which is between 1 and 4
+    >>> fun(4)
+    SlycotArithmeticError:
+    4 is in [4,    1.2e+02)!
+    >>> fun(120)
+    SlycotResultWarning:
+    120 is too large
+    """
+    if docstring:
```

Comment:
shouldn't we check info > 0?
Now we are parsing the docstring after each calculation.

Reply:
Custom messages with info < 0 should be possible

Reply:
ok, but I am worried about performance
can we do
```python
    if docstring and (iwarn or info):
        ....
```
then we don't parse the docstring upon succesful calculation. Might matter to someone who is calculating a lot of LQR controllers or the like.